sdk/servicebus/Azure.Messaging.ServiceBus/src/Authorization/SharedAccessCredential.cs

```diff
@@ -129,17 +128,17 @@ public SharedAccessCredential(AzureSasCredential sourceCredential)
                 if ((!string.Equals(signature.SharedAccessKeyName, name, StringComparison.Ordinal))
                     || (!string.Equals(signature.SharedAccessKey, key, StringComparison.Ordinal)))
                 {
-                    signature = new SharedAccessSignature(signature.Resource, name, key);
-                    Volatile.Write(ref _sharedAccessSignature, signature);
+                    var updatedSignature = new SharedAccessSignature(signature.Resource, name, key);
+                    signature = SafeUpdateSharedAccessSignature(signature, updatedSignature);
```

Comment:
Did you consider creating a new SharedAccessSignature for each call to avoid having to correlate the key credential with the SasSignature? I believe this would only get called once per 45 minutes or so.

Reply:
We could potentially go down that path, as the normal refresh for an open connection/link would be every 25 minutes for a shared key.  It may be more frequent for a SAS, as we don't control the expiration time.  However, we do need this value any time a new connection or link is opened as well, including on recovery.  So, every 20 minutes or so when the service force-closes connections, we would need potentially need the value several times in a short window as we reconnect and reopen links.   Likewise if we see anything idle-out or get closed by exception, we'd need it more frequent.

There's also the service consideration; if we were to regenerate on each call, we'd be giving the service a distinct token (due to different expiration time) for each connection/link that is opened.  I'd prefer not to change the client/server interaction without talking to the service teams to understand if that would put additional pressure on the service or cause them concern for another reason.

I'm going to keep this as-is for now;  we've already got the code written and tested.   The caching does save us from allocating repeatedly for the same value when we're opening things, and it keeps from changing the client behavior when talking to the service.